tests/test_worker.py

```diff
@@ -17,39 +18,63 @@ def handle(self, task):
         self.test_done += 1
 
 
-class WorkerTest(TestCase):
-    def test_run(self):
-        conn = get_fakeredis()
-        operation = "lala"
-        worker = CountingWorker(conn=conn, stages=[operation])
-        worker.sync()
-        assert worker.test_done == 0, worker.test_done
-        job = Job.create(conn, "test")
-        stage = job.get_stage(operation)
-        task = stage.queue({}, {})
-        assert not job.is_done()
-        assert worker.test_done == 0, worker.test_done
-        worker.sync()
-        assert worker.test_done == 1, worker.test_done
-        assert job.is_done()
-        worker.retry(task)
-        assert not job.is_done()
-        worker.sync()
-        assert job.is_done()
-        assert worker.exit_code == 0, worker.exit_code
-        assert worker.test_done == 1, worker.test_done
-        worker.retry(task)
-        worker.run(blocking=False)
-        assert job.is_done()
-        assert worker.exit_code == 0, worker.exit_code
-        worker.num_threads = None
+class NoOpWorker(worker.Worker):
+    def handle(self, task):
+        pass
+
+
+def test_run():
+    conn = get_fakeredis()
+    operation = "lala"
+    worker = CountingWorker(conn=conn, stages=[operation])
+    worker.sync()
+    assert worker.test_done == 0, worker.test_done
+    job = Job.create(conn, "test")
+    stage = job.get_stage(operation)
+    task = stage.queue({}, {})
+    assert not job.is_done()
+    assert worker.test_done == 0, worker.test_done
+    worker.sync()
+    assert worker.test_done == 1, worker.test_done
+    assert job.is_done()
+    worker.retry(task)
+    assert not job.is_done()
+    worker.sync()
+    assert job.is_done()
+    assert worker.exit_code == 0, worker.exit_code
+    assert worker.test_done == 1, worker.test_done
+    worker.retry(task)
+    worker.run(blocking=False)
+    assert job.is_done()
+    assert worker.exit_code == 0, worker.exit_code
+    worker.num_threads = None
+    worker.retry(task)
+    worker.run(blocking=False)
+    assert job.is_done()
+    assert worker.exit_code == 0, worker.exit_code
+    try:
+        worker._handle_signal(5, None)
+    except SystemExit as exc:
+        assert exc.code == 5, exc.code
+    with pytest.raises(SystemExit) as exc:  # noqa
+        worker._handle_signal(5, None)
+
+
+def test_fails(caplog):
+    caplog.set_level(logging.DEBUG)
+    conn = get_fakeredis()
+    operation = "fails"
+    job = Job.create(conn, "test")
+    stage = job.get_stage(operation)
+    task = stage.queue({}, {})
+
+    worker = NoOpWorker(conn=conn, stages=[operation])
+    worker.sync()
+    for _ in range(4):
         worker.retry(task)
```

Comment:
I’m not sure if it would work, but I think something like this could be a bit closer to a real world scenario and depending less on implementation details:

```py
for _ in range(4):
  worker.sync()
```

Reply:
You are right, but I couldn't immediately get it to work. I will play around with this some more.

Reply:
If you can’t get it to work within a reasonable amount of time, feel free to ignore it. The way the test is implemented currently is fine, too!